tls/s2n_psk.c

```diff
@@ -289,3 +289,55 @@ S2N_RESULT s2n_finish_psk_extension(struct s2n_connection *conn)
     GUARD_RESULT(s2n_psk_write_binder_list(conn, &partial_client_hello, client_hello));
     return S2N_RESULT_OK;
 }
+
+static int s2n_psk_set_hmac(struct s2n_psk *psk, s2n_psk_hmac psk_hmac_alg)
+{
+    switch(psk_hmac_alg) {
+        case S2N_PSK_HMAC_SHA224:     psk->hmac_alg = S2N_HMAC_SHA224; break;
+        case S2N_PSK_HMAC_SHA256:     psk->hmac_alg = S2N_HMAC_SHA256; break;
+        case S2N_PSK_HMAC_SHA384:     psk->hmac_alg = S2N_HMAC_SHA384; break;
+        default:
+            S2N_ERROR(S2N_ERR_HMAC_INVALID_ALGORITHM);
+    }
+    return S2N_SUCCESS;
+}
+
+int s2n_connection_set_external_psks(struct s2n_connection *conn, struct s2n_external_psk *psk_vec, size_t psk_vec_length)
+{
+    notnull_check(conn);
+    notnull_check(psk_vec);
+    
+    /* Remove all previously-set external psks */
+    for (size_t i = conn->psk_params.psk_list.len; i > 0; i--) {
+        /* This index variable along with the loop condition keep size_t i from underflowing */
+        size_t index = i - 1;
+        struct s2n_psk *psk = NULL;
+        GUARD_AS_POSIX(s2n_array_get(&conn->psk_params.psk_list, index, (void**) &psk));
+        ENSURE_POSIX_REF(psk);
```

Comment:
Nit: You probably shouldn't used notnull_check() and ENSURE_POSIX_REF() in the same method. I'd pick one-- probably notnull_check, which is used more widely.